File: general-python-scripts/Americans for the Arts Local Index Pull.py
#-------------------------------------------------------------------------------
# Name:        Americans for the Arts Local Index Extractor
# Purpose:     This pulls the data from the AftA Local Index into a csv file
#
# Author:      Michael Silva
#
# Created:     20/11/2014
#-------------------------------------------------------------------------------

import requests
import pandas
import os

# This function creates a text progress bar
def progressBar(i, j, size_of_pb = 40):
    percent_done = i / j
    pb_done = round(size_of_pb * percent_done, 0)
    pb_left = size_of_pb - pb_done
    percent_done = round(percent_done*100)
    return('|' + '='*int(pb_done) + '>' + ' '*int(pb_left) + '| ' +str(percent_done)+'%')

# Initialize Progress Bar
pb_i=0
print(progressBar(pb_i, len(states)))

states = ['AK','AL','AR','AZ','CA','CO','CT','DC','DE','FL','GA','HI','IA','ID','IL','IN','KS','KY','LA','MA','MD','ME','MI','MN','MO','MS','MT','NC','ND','NE','NH','NJ','NM','NV','NY','OH','OK','OR','PA','RI','SC','SD','TN','TX','UT','VA','VT','WA','WI','WV','WY']

# Step through the states
for st in states:
    print('Starting '+st)
    # Get the counties in the state
    url1 = 'http://www.artsindexusa.org/fetchCounties.php?state='+st
    r1 = requests.get(url1)
    json1 = r1.json()
    # Step through the counties
    for fips in json1:
        s = fips.split(':')
        county_fips = s[0]
        county_name = s[1]
        #print(county_name,st) # Useful for debugging
        url2 = 'http://www.artsindexusa.org/fetchCounty.php?selectedCounty='+county_fips
        df = pandas.io.json.read_json(url2)
        df['row'] = df.index.values
        df['fips'] = county_fips
        df['county name'] = county_name
        df['state'] = st

        # Store the data in a data frame
        try:
            arts_data
        except NameError:
            # The data frame does not exist so let's create it
            arts_data = df
        else:
            # Append the rows to the data frame
            arts_data = arts_data.append(df)

    # Update progress bar
    pb_i = pb_i + 1
    print(progressBar(pb_i, len(states)))

# Save the Files to the Desktop
userhome = os.path.expanduser('~')
desktop = userhome + '/Desktop/'
file_path = desktop + 'Local Arts Index.csv'
arts_data.to_csv(file_path, sep=',', encoding='utf-8')

File: general-r-scripts/Americans for the Arts Local Index Pull.R
## This script downloads the data from Americans for the Arts
## Local Arts Index (http://www.artsindexusa.org/local-arts-index)

library(jsonlite)

base.url <- 'http://www.artsindexusa.org/fetchCounty.php?selectedCounty='

states <- c('AK','AL','AR','AZ','CA','CO','CT','DC','DE','FL','GA','HI','IA','ID','IL','IN','KS','KY','LA','MA','MD','ME','MI','MN','MO','MS','MT','NC','ND','NE','NH','NJ','NM','NV','NY','OH','OK','OR','PA','RI','SC','SD','TN','TX','UT','VA','VT','WA','WI','WV','WY')

## Progress bar
pb <- txtProgressBar(min=0, max=length(states), style=3)
pb.i <- 0

## Step through each state
for(st in states){
  message('Starting ',st)
  
  ## Get county list
  url <- paste0('http://www.artsindexusa.org/fetchCounties.php?state=', st)
  county.fips <- fromJSON(url)
  
  ## Step through each county
  for(fips in county.fips){
    s <- unlist(strsplit(fips,":"))[1]
    county.name <- unlist(strsplit(fips,":"))[2]
    url <- paste0(base.url, s)
    json.data <- as.data.frame(fromJSON(url))
    json.data$row <- 1:nrow(json.data)
    json.data$fips <- s
    json.data$county_name <- county.name
    json.data$state <- st
    ## Store data
    if(exists("arts.data")){
      arts.data <- rbind(arts.data, json.data)
    } else{
      arts.data <- json.data
    }
  }
  
  ## Update Progress Bar
  pb.i <- pb.i + 1
  setTxtProgressBar(pb, pb.i)
}
close(pb)

write.csv(arts.data,'Local Arts Index.csv')